packages/@aws-cdk/aws-iot-actions/lib/s3-action.ts

```diff
@@ -0,0 +1,77 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as s3 from '@aws-cdk/aws-s3';
+import { kebab as toKebabCase } from 'case';
+import { singletonActionRole } from './private/role';
+
+/**
+ * Configuration properties of an action for s3.
+ */
+export interface S3ActionProps {
+  /**
+   * The Amazon S3 canned ACL that controls access to the object identified by the object key.
+   * @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html#canned-acl
+   *
+   * @default None
+   */
+  readonly cannedAcl?: s3.BucketAccessControl;
+
+  /**
+   * The path to the file where the data is written.
+   *
+   * Supports substitution templates.
+   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-substitution-templates.html
+   *
+   * @default '${topic()}/${timestamp()}'
+   */
+  readonly key?: string;
+
+  /**
+   * The IAM role that allows access to the S3.
+   *
+   * @default a new role will be created
+   */
+  readonly role?: iam.IRole;
```

Comment:
Will every Action have a `role` property? Because if the answer is "yes", we should introduce a common interface that all Action prop interfaces extend. Like the `CommonActionProps` in `@aws-cdk/aws-codepipeline`.

Reply:
@skinny85 
Not every Action is need a role, but almost. 2 actions in [all actions](https://docs.aws.amazon.com/iot/latest/developerguide/iot-rule-actions.html) will not have a `role` property. 

- HTTP Action
- Lambda Action

For example, `LambdaFunctionAction` constructor (most simple one) is as following:
```ts
constructor(private readonly func: lambda.IFunction) {}
```

So I thought there are no common property for all actions. But the way to align the arguments of the action's constructor is what I was struggling with.

What do you think? I'm open to any and all refactoring!

Reply:
Ok, so if all of actions besides two will need `role`, I think introducing a separate common interface makes sense. And make `S3ActionProps` (soon to be changed to `S3PutObjectActionProps` I hope 😉) extend this new interface.

Reply:
OK! I will! 🚀

Reply:
@skinny85 
If we create `CommonActionProps` and `CommonAwsActionProps`, should these constructors be unified in the following form?

```ts
interface CommonActionProps {
}

intercace CommonAwsActionProps extends CommonActionProps {
  role?: iam.IRole;
}

class Action {
  constructor(props: CommonActionProps) {}
}
```

Now, three Actions `LambdaFunctionAction`, `CloudWatchLogsAction` and `S3PutObjectActionProps` have following constructor, and first argument of them constroctor is not `Props`:

`LambdaFunctionAction`
```ts
constructor(func: lambda.IFunction) {}
```

`CloudWatchLogsAction`
```ts
constructor(logGroup: logs.ILogGroup, props: CloudWatchLogsActionProps) {}
```

`S3PutObjectActionProps`
```ts
constructor(bucket: s3.IBucket, props: S3PutObjectActionProps) {}
```

So if first question is "yes", I will do this refactoring in another PR!

Reply:
No. Let's leave the constructors as they are now.

BTW, there's no point in having both `CommonActionProps` and `CommonAwsActionProps` for IoT Actions - you only need one common superinterface (it's a different story with CodePipeline Actions, but I don't want to get into the details of why that is here).